js/app/downloader.js

```diff
@@ -36,74 +35,33 @@ const FileDownloader = new Lang.Class({
         Soup.Session.prototype.add_feature.call(this._http_session, this._requester);
     },
 
-    verify_checksum: function (file, against_digest, cancellable, callback) {
-        let task = new AsyncTask.AsyncTask(this, cancellable, callback);
-        task.catch_errors(() => {
+    verify_checksum_promise: function (file, against_digest, cancellable) {
+        return Promise.resolve().then(() => {
             let stream = file.read(cancellable);
             let csum = new GLib.Checksum(GLib.ChecksumType.SHA256);
 
-            function finish() {
-                let digest = csum.get_string();
-                if (against_digest.toLowerCase() !== digest.toLowerCase())
-                    throw new Error(Format.vprintf("Checksum against file %s did not match: %s (local), %s (remote)", [file.get_basename(), digest, against_digest]));
-                task.return_value(true);
-            }
-
-            function bytes_read(stream, res) {
-                let bytes = stream.read_bytes_finish(res);
-                csum.update(bytes.get_data());
-
-                if (bytes.get_size() !== 0)
-                    continue_checksum();
-                else
-                    finish();
-            }
+            function continue_checksum () {
```

Comment:
We do still need to recursively call into it until we hit the end of our stream. Though if you see another way to unroll it let me know.